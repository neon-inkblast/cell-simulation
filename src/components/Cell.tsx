import styled from "styled-components";
import React from "react";
import { useCellGridContext } from "../context/CellGridContext";

export type CellProps = {
  x?: number;
  y?: number;
  alive?: boolean;
};

const StyledCell = styled.div`
  width: 50px;
  height: 50px;
  border-radius: 20px;
  border: 1px solid black;
  display: inline-block;
  margin: 0;
  background-color: ${({ alive }: CellProps) => (alive ? "lightseagreen" : "transparent")};

  &:hover {
    background-color: ${({ alive }: CellProps) => (alive ? "gold" : "yellow")};
  }
`;

const Cell: React.FC<CellProps> = ({ y, x, alive }) => {
  const { dispatch } = useCellGridContext();
  return <StyledCell alive={alive} onClick={() => dispatch({ type: "toggle-cell", column: x, row: y })}></StyledCell>;
};

export default Cell;
