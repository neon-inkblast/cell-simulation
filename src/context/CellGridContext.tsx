import React, { useReducer } from "react";
import { CellGridAction, CellGridContextType, CellGridState } from "../@types/CellGridContext";
import { GridConfig } from "../config/gridConfig";
import { generateNext, initialState, toggleCell } from "./cellGridUtils";

export const CellGridContext = React.createContext<CellGridContextType | undefined>(undefined);

// reducer for the state in this context
function gridReducer(state: CellGridState, action: CellGridAction): CellGridState {
  switch (action.type) {
    case "reset": {
      return initialState(GridConfig.rows, GridConfig.columns);
    }
    case "toggle-cell": {
      return toggleCell(state, action.row as number, action.column as number);
    }
    case "generate-next": {
      return generateNext(state);
    }
    default:
      throw Error(`Action type "${action.type}" not defined.  No state change`);
  }
}

// reducer wrapped provider component
export function CellGridProvider({ children }: { children: React.ReactNode }) {
  const [state, dispatch] = useReducer(gridReducer, initialState(GridConfig.rows, GridConfig.columns));
  const value = { state, dispatch };
  return <CellGridContext.Provider value={value}> {children} </CellGridContext.Provider>;
}

// hook around useContext to show an error if used outside provider
export function useCellGridContext(): CellGridContextType {
  const context = React.useContext(CellGridContext);
  if (context === undefined) {
    throw new Error("useGrid must be used within a GridProvider");
  }
  return context;
}

export default { CellGridContext, CellGridProvider, useCellGridContext };
