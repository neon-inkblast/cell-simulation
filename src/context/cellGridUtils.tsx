import { CellGridState } from "../@types/CellGridContext";
/*
 * Utility functions to perform the state transforms for the CellGridProvider
 */
export const generateNext = (state: CellGridState): CellGridState => {
  const { grid: prevGrid, generation } = state;
  // calculate max rows and columns, for cell neighbour wrapping
  const lastRowIndex = prevGrid.length ? prevGrid.length - 1 : 0;
  const lastColIndex = prevGrid[0]?.length ? prevGrid[0].length - 1 : 0;

  // inner function to use when mapping current cells to next gen cells
  function getNextCellState(x: number, y: number, cell: boolean): boolean {
    // calculate surrounding rows and columns, taking wrapping into account
    const prevRow = y === 0 ? lastRowIndex : y - 1;
    const nextRow = y === lastColIndex ? 0 : y + 1;
    const prevCol = x === 0 ? lastRowIndex : x - 1;
    const nextCol = x === lastColIndex ? 0 : x + 1;

    // get an array of the neighbouring cells
    const neighbours = [
      prevGrid[prevRow][prevCol],
      prevGrid[prevRow][x],
      prevGrid[prevRow][nextCol],
      prevGrid[y][prevCol],
      prevGrid[y][nextCol],
      prevGrid[nextRow][prevCol],
      prevGrid[nextRow][x],
      prevGrid[nextRow][nextCol],
    ];

    // count the "live" neighbouring cells
    const liveNeighbours = neighbours.reduce((acc, current) => (acc += current ? 1 : 0), 0);
    // if live neighbours is 3, become or stay alive
    // otherwise, if 2 and alive, stay alive
    return liveNeighbours === 3 || (liveNeighbours === 2 && cell);
  }

  // map the new cell states from the old cells
  const grid = prevGrid.map((row, y) =>
    row.map((cell, x) => {
      return getNextCellState(x, y, cell);
    })
  );

  return {
    generation: generation + 1,
    grid,
  };
};

export const toggleCell = (state: CellGridState, cellY: number, cellX: number): CellGridState => {
  const prevGrid = state.grid;
  const grid = prevGrid.map((row, y) => {
    // if not the target row, skip any more checks and just recreate the same row
    if (y !== cellY) {
      return [...row];
    }
    // if in the target row, toggle when the x position matches
    return row.map((cell, x) => {
      if (cellX === x) {
        return !cell;
      }
      return cell;
    });
  });
  return {
    ...state,
    grid,
  };
};

export const initialState = (rows = 5, columns = 5): CellGridState => {
  return {
    generation: 0,
    grid: Array(rows).fill(Array(columns).fill(false)),
  };
};

const cellGridUtils = { generateNext, toggleCell, initialState };

export default cellGridUtils;
