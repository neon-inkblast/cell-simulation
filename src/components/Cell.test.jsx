import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { CellGridContext } from "../context/CellGridContext";
import Cell from "./Cell";

const customRender = (ui, { providerProps, ...renderOptions }) => {
  return render(<CellGridContext.Provider value={{ ...providerProps }}>{ui}</CellGridContext.Provider>, renderOptions);
};

describe("Cell component", () => {
  const testX = 1;
  const testY = 2;
  let renderResult;
  let providerProps;

  beforeEach(() => {
    providerProps = {
      state: [
        [0, 1],
        [1, 0],
      ],
      dispatch: jest.fn(),
    };
    renderResult = customRender(<Cell x={testX} y={testY} alive={true} />, { providerProps });
  });

  it("renders", () => {
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("dispatches a toggle cell event on click, with x/y props for context state", () => {
    // Click the cell
    userEvent.click(renderResult.container.firstChild);
    // Verify dispatch to context state was called
    expect(providerProps.dispatch).toHaveBeenCalledTimes(1);
    expect(providerProps.dispatch).toHaveBeenCalledWith({ type: "toggle-cell", column: testX, row: testY });
  });
});
