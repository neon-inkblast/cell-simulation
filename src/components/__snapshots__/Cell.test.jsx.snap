// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell component renders 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBQB jhQtaX"
  />
</DocumentFragment>
`;
